Update command injection docs (#741)
returntocorp/semgrep-docs: docs/cheat-sheets/go-command-injection.mdx

---

import LinkToRegistryRule from "/src/components/LinkToRegistryRule"
import CommandInjectionIntro from "/src/components/concept/_command-injection-intro.mdx"

# Command injection prevention for Go

<CommandInjectionIntro />

#### Check your project for these conditions:

```bash
semgrep scan --config p/go-command-injection
```

---

## 1. Running an OS command

### 1.A. running OS commands with exec.Command()

> The `Command` and `CommandContext` return the Cmd struct to execute the named program with the given arguments. If unverified user data can reach its call site, this may end up in a command injection vulnerability. Both `Command` and `CommandContext` have built-in protections that do not let command arguments cause trouble. But ensure that the command itself is not controlled by the user, also do not use `sh`, because internal protection does not work in this case.

Author: ```suggestion
The `Command` and `CommandContext` execute commands with provided arguments. If unverified user data can reach its call site, this may end up in a command injection vulnerability. Both `Command` and `CommandContext` have built-in protections that do not let command arguments cause trouble. But ensure that the command itself is not controlled by the user, also do not use `sh`, because internal protection does not work in this case.
```

  // Command example
  cmd := exec.Command("echo", "hello")

  // CommandContext example
  err := exec.CommandContext(ctx, "sleep", "5").Run()

  // Not vulnerable
  cmd := exec.Command("echo", "1; cat /etc/passwd")

  // Vunerable
  userInput :=  "echo 1 | cat /etc/passwd" // value supplied by user input
  out, _ = exec.Command("sh", "-c", userInput).Output()

  // Vulnerable
  userInput1 := "cat" // value supplied by user input
  userInput2 := "/etc/passwd" // value supplied by user input
  out, _ = exec.Command(userInput1, userInput2)
}
```

#### References

- [`Command` documentation](https://golang.org/pkg/os/exec/#Command)
- [`CommandContext` documentation](https://golang.org/pkg/os/exec/#CommandContext)

#### Mitigation

Do not let user input in `exec.Command` and `exec.CommandContext` functions. Alternatively:

> - Always try to use internal Go API (if it exists) instead of running an OS command.

Author: ```suggestion
- Always try to use internal Go API (if it exists) instead of running an OS command. In other words, use internal language features instead of invoking commands that can be exploited.
- Avoid running a shell as a command with arguments, if it is not possible, then strip the input of everything except alphanumeric characters provided for the command string and arguments.
```

- Try to avoid non-literal values for the command string.
- If it is not possible, then do not let running arbitrary commands, use an allowlist for inputs.

> - Avoid running `sh` as a command with arguments, if it is not possible - strip everything except alphanumeric characters from an input provided for the command string and arguments.

Author: ```suggestion
```

  // Command example
  cmd := exec.Command("echo", "hello")

  // CommandContext example
  err := exec.CommandContext(ctx, "sleep", "5").Run()

  // Not vulnerable
  cmd := exec.Command("echo", "1; cat /etc/passwd")

  // Vunerable
  userInput :=  "echo 1 | cat /etc/passwd" // value supplied by user input
  out, _ = exec.Command("sh", "-c", userInput).Output()

  // Vulnerable
  userInput1 := "cat" // value supplied by user input
  userInput2 := "/etc/passwd" // value supplied by user input
  out, _ = exec.Command(userInput1, userInput2)
}
```

#### References

- [`Command` documentation](https://golang.org/pkg/os/exec/#Command)
- [`CommandContext` documentation](https://golang.org/pkg/os/exec/#CommandContext)

#### Mitigation

Do not let user input in `exec.Command` and `exec.CommandContext` functions. Alternatively:

- Always try to use internal Go API (if it exists) instead of running an OS command. In other words, use internal language features instead of invoking commands that can be exploited.

> - Avoid running shell as a command with arguments, if it is not possible, then strip the input of everything except alphanumeric characters provided for the command string and arguments.

Author: ```suggestion
- Avoid running shell as a command with arguments, if it is not possible, then strip everything except alphanumeric characters from an input provided for the command string and arguments.
```

func main() {
  cmd := &exec.Cmd {
    // Path is the path of the command to run.
    Path: "/bin/echo",
    // Args holds command line arguments, including the command itself as Args[0].
    Args: []string{ "echo", "hello world" },
    Stdout: os.Stdout,
    Stderr: os.Stdout,
  }
  cmd.Start();
  cmd.Wait();

  // Args also can be ommited and {Path} will be used by default
  cmd := &exec.Cmd {
    Path: "/bin/echo"
  }

  // Vulnerable
  userInput := "/pwn/exploit" // value supplied by user input
  cmd := &exec.Cmd {
    Path: userInput
  }

  // Vulnerable
  userInput1 := "/bin/bash" // value supplied by user input
  userInput2 := []string{ "bash", "exploit.sh" } // value supplied by user input
  cmd := &exec.Cmd {
    Path: userInput1,
    Args: userInput2
  }

}
```

#### References

- [`Cmd` struct documentation](https://golang.org/pkg/os/exec/#Cmd)

#### Mitigation

Do not let user input in `exec.Cmd` struct. Alternatively:

- Always try to use internal Go API (if it exists) instead of running an OS command.
- Try to avoid non-literal values for the command string.

> - If it is not possible, then do not let running arbitrary commands, use a white list for inputs.

Author: ```suggestion
- If it is not possible, then do not let running arbitrary commands, use an allow list for inputs.
```
